aws/external/resolve_credentials.go

```diff
@@ -0,0 +1,404 @@
+				options.AuthorizationToken = authToken
+			}
+		},
+	}
+
+	options, found, err := GetEndpointCredentialProviderOptions(configs)
+	if err != nil {
+		return err
+	}
+	if found {
+		opts = append(opts, options)
+	}
+
+	provider := endpointcreds.New(cfgCopy, opts...)
+
+	cfg.Credentials = provider
+
+	return nil
+}
+
+func resolveEC2RoleCredentials(cfg *aws.Config, configs Configs) error {
+	cfgCpy := *cfg
+
+	opts := []func(*ec2rolecreds.ProviderOptions){
+		func(options *ec2rolecreds.ProviderOptions) {
+			options.ExpiryWindow = 5 * time.Minute
+		},
+	}
+
+	options, found, err := GetEC2RoleCredentialProviderOptions(configs)
+	if err != nil {
+		return err
+	}
+	if found {
+		opts = append(opts, options)
+	}
+
+	provider := ec2rolecreds.New(ec2metadata.New(cfgCpy), opts...)
+
+	cfg.Credentials = provider
+
+	return nil
+}
+
+func resolveCredsFromSource(cfg *aws.Config, envConfig *EnvConfig, sharedCfg *SharedConfig, configs Configs) (err error) {
+	switch sharedCfg.CredentialSource {
+	case credSourceEc2Metadata:
+		err = resolveEC2RoleCredentials(cfg, configs)
+	case credSourceEnvironment:
+		cfg.Credentials = aws.StaticCredentialsProvider{Value: envConfig.Credentials}
+	case credSourceECSContainer:
+		if len(envConfig.ContainerCredentialsRelativePath) == 0 {
+			return ErrSharedConfigECSContainerEnvVarEmpty
+		}
+		return resolveHTTPCredProvider(cfg, ecsContainerURI(envConfig.ContainerCredentialsRelativePath), envConfig.ContainerAuthorizationToken, nil)
+	default:
+		return ErrSharedConfigInvalidCredSource
+	}
+
+	return nil
+}
+
+func assumeWebIdentity(cfg *aws.Config, filepath string, roleARN, sessionName string, configs Configs) error {
+	if len(filepath) == 0 {
+		return WebIdentityEmptyTokenFilePathErr
+	}
+
+	if len(roleARN) == 0 {
+		return WebIdentityEmptyRoleARNErr
+	}
+
+	var opts []func(*stscreds.WebIdentityRoleProviderOptions)
+
+	options, found, err := GetWebIdentityCredentialProviderOptions(configs)
+	if err != nil {
+		return err
+	}
+	if found {
+		opts = append(opts, options)
+	}
+
+	provider := stscreds.NewWebIdentityRoleProvider(sts.New(*cfg), roleARN, sessionName, filepath, opts...)
+
+	cfg.Credentials = provider
+
+	return nil
+}
+
+func credsFromAssumeRole(cfg *aws.Config, sharedCfg *SharedConfig, configs Configs) (err error) {
+	var tokenFunc func() (string, error)
+	if len(sharedCfg.MFASerial) != 0 {
+		var found bool
+		tokenFunc, found, err = GetMFATokenFunc(configs)
+		if err != nil {
+			return err
+		}
+
+		if !found {
+			// AssumeRole Token provider is required if doing Assume Role
+			// with MFA.
+			return AssumeRoleTokenProviderNotSetError{}
+		}
+	}
+
+	sts := sts.New(*cfg)
+
+	opts := []func(*stscreds.AssumeRoleProviderOptions){
+		func(options *stscreds.AssumeRoleProviderOptions) {
+			options.RoleSessionName = sharedCfg.RoleSessionName
+			if sharedCfg.RoleDurationSeconds != nil {
+				if *sharedCfg.RoleDurationSeconds/time.Second < 900 {
+					options.Duration = time.Second * 900
```

Comment:
Minor nit but using `time.Minute*15` would be a bit easier to read then 900. (Don't have to think about how many minutes 900 seconds are)